core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java

```diff
@@ -604,16 +608,37 @@ public Object visitRemoveExpiredCommand(InvocationContext ctx, RemoveExpiredComm
 
    @Override
    public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throws Throwable {
-      ClusterPublisherManager managerToUse;
-      if (command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-         managerToUse = localClusterPublisherManager;
-      } else {
-         managerToUse = clusterPublisherManager;
+      CompletionStage<Long> sizeStage = trySizeOptimization(command.getFlagsBitSet())
+            .thenCompose(size -> {
+               if (size > -1) {
+                  return CompletableFuture.completedFuture(size);
+               }
+
+               ClusterPublisherManager<?, ?> managerToUse;
+               if (command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
+                  managerToUse = localClusterPublisherManager;
+               } else {
+                  managerToUse = clusterPublisherManager;
+               }
+
+               return managerToUse.keyReduction(false, null, null, ctx,
+                     command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(),
+                     PublisherReducers.add());
+            });
+      return asyncValue(sizeStage);
+   }
+
+   private CompletionStage<Long> trySizeOptimization(long flags) {
+      if (!EnumUtil.containsAny(flags, FlagBitSets.CACHE_MODE_LOCAL)
+            || EnumUtil.containsAny(flags, FlagBitSets.SKIP_CACHE_LOAD | FlagBitSets.SKIP_SIZE_OPTIMIZATION)
```

Comment:
Why does using `SKIP_CACHE_LOAD` preclude this optimization? I wouldn't expect to test that here at all.